File: web_api/contracts.py
"""Pydantic contracts for the read-only TRAIDR web API."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Generic, Literal, TypeVar

from pydantic import BaseModel, ConfigDict, Field

ApiStatus = Literal["OK", "DEGRADED", "NO_TRADE", "INSUFFICIENT_DATA", "ERROR"]
FreshnessState = Literal["FRESH", "STALE", "UNKNOWN"]


class ReadModel(BaseModel):
    """Base model that rejects undeclared response fields."""

    model_config = ConfigDict(extra="forbid")


class FreshnessInfo(ReadModel):
    state: FreshnessState
    observed_at: datetime | None = None
    age_seconds: float | None = Field(default=None, ge=0)
    maximum_age_seconds: float | None = Field(default=None, gt=0)


ModelT = TypeVar("ModelT")


class ApiEnvelope(ReadModel, Generic[ModelT]):
    status: ApiStatus
    as_of: datetime
    data: ModelT
    freshness: dict[str, FreshnessInfo] = Field(default_factory=dict)
    reason_codes: tuple[str, ...] = ()
    can_execute_trades: Literal[False] = False
    request_id: str = Field(min_length=1, max_length=128)


class StatusData(ReadModel):
    database_exists: bool
    table_count: int = Field(ge=0)
    service_state: str
    latest_heartbeat: datetime | None = None
    provider_health: list[dict[str, Any]] = Field(default_factory=list)
    safety: dict[str, Any] = Field(default_factory=dict)


class ShadowEvidenceData(ReadModel):
    provider: str
    observed_at: datetime | None = None
    freshness: FreshnessState = "UNKNOWN"
    setup_class: str = "INSUFFICIENT_DATA"
    market_regime: str = "INSUFFICIENT_DATA"
    crowding_score: float | None = None
    squeeze_risk: float | None = None
    catalyst_risk: float | None = None
    data_quality_score: float = 0.0
    probability_state: str = "UNCALIBRATED"
    scoring_weight: float = Field(default=0.0, ge=0.0, le=0.0)
    fields: dict[str, float] = Field(default_factory=dict)
    conflicts: list[str] = Field(default_factory=list)
    reason_codes: list[str] = Field(default_factory=list)
    can_execute_trades: Literal[False] = False


class ScannerFactorData(ReadModel):
    factor: str
    weight: float
    raw_value: float | None = None
    normalized_value: float | None = None
    long_contribution: float = 0.0
    short_contribution: float = 0.0
    explanation: str
    source: str | None = None
    reason_codes: list[str] = Field(default_factory=list)


class ScannerRow(ReadModel):
    score_id: str
    instrument_id: str
    observed_at: datetime | None = None
    recorded_at: datetime | None = None
    status: str
    direction: str
    score: float | None = None
    long_score: float | None = None
    short_score: float | None = None
    risk_score: float | None = None
    conflicts: list[str] = Field(default_factory=list)
    reason_codes: list[str] = Field(default_factory=list)
    factors: list[ScannerFactorData] = Field(default_factory=list)
    shadow: ShadowEvidenceData | None = None
    can_execute_trades: Literal[False] = False


class ScannerData(ReadModel):
    rows: list[ScannerRow] = Field(default_factory=list)
    limit: int = Field(ge=1, le=2_000)
    total: int = Field(ge=0)


class OverviewData(ReadModel):
    scanner: list[ScannerRow] = Field(default_factory=list)
    alerts: list[dict[str, Any]] = Field(default_factory=list)
    paper_positions: list[dict[str, Any]] = Field(default_factory=list)
    service_heartbeats: list[dict[str, Any]] = Field(default_factory=list)
    shadow_evidence: list[ShadowEvidenceData] = Field(default_factory=list)
    news: list[dict[str, Any]] = Field(default_factory=list)


class CandleData(ReadModel):
    open_time_ms: int
    open: float
    high: float
    low: float
    close: float
    base_volume: float
    quote_volume: float
    source: str
    received_at: datetime
    quality: str
    can_execute_trades: Literal[False] = False


class MarketData(ReadModel):
    instrument_id: str
    interval: str
    candles: list[CandleData] = Field(default_factory=list)
    scanner: ScannerRow | None = None
    microstructure: list[dict[str, Any]] = Field(default_factory=list)
    evidence: list[dict[str, Any]] = Field(default_factory=list)


class MarketInstrumentData(ReadModel):
    instrument_id: str
    source: str
    symbol: str
    base_asset: str
    quote_asset: str
    status: str
    discovered_at: datetime
    refreshed_at: datetime
    can_execute_trades: Literal[False] = False


class MarketsData(ReadModel):
    instruments: list[MarketInstrumentData] = Field(default_factory=list)
    limit: int = Field(ge=1, le=2_000)
    total: int = Field(ge=0)


class ErrorData(ReadModel):
    code: str
    message: str

File: web/src/lib/contracts.ts
export type ApiStatus = "OK" | "DEGRADED" | "NO_TRADE" | "INSUFFICIENT_DATA" | "ERROR";
export type FreshnessState = "FRESH" | "STALE" | "UNKNOWN";
export type DataSource = "local-api" | "preview";

export interface FreshnessInfo {
  state: FreshnessState;
  observed_at: string | null;
  age_seconds: number | null;
  maximum_age_seconds: number | null;
}

export interface ApiEnvelope<T> {
  status: ApiStatus;
  as_of: string;
  data: T;
  freshness: Record<string, FreshnessInfo>;
  reason_codes: string[];
  can_execute_trades: false;
  request_id: string;
}

export interface StatusData {
  database_exists: boolean;
  table_count: number;
  service_state: string;
  latest_heartbeat: string | null;
  provider_health: Array<Record<string, unknown>>;
  safety: Record<string, unknown>;
}

export interface ShadowEvidence {
  provider: string;
  observed_at: string | null;
  freshness: FreshnessState;
  setup_class: string;
  market_regime: string;
  crowding_score: number | null;
  squeeze_risk: number | null;
  catalyst_risk: number | null;
  data_quality_score: number;
  probability_state: string;
  scoring_weight: 0;
  fields: Record<string, number>;
  conflicts: string[];
  reason_codes: string[];
  can_execute_trades: false;
}

export interface ScannerFactor {
  factor: string;
  weight: number;
  raw_value: number | null;
  normalized_value: number | null;
  long_contribution: number;
  short_contribution: number;
  explanation: string;
  source: string | null;
  reason_codes: string[];
}

export interface ScannerRow {
  score_id: string;
  instrument_id: string;
  observed_at: string | null;
  recorded_at: string | null;
  status: string;
  direction: string;
  score: number | null;
  long_score: number | null;
  short_score: number | null;
  risk_score: number | null;
  conflicts: string[];
  reason_codes: string[];
  factors: ScannerFactor[];
  shadow: ShadowEvidence | null;
  can_execute_trades: false;
}

export interface ScannerData {
  rows: ScannerRow[];
  limit: number;
  total: number;
}

export interface OverviewData {
  scanner: ScannerRow[];
  alerts: Array<Record<string, unknown>>;
  paper_positions: Array<Record<string, unknown>>;
  service_heartbeats: Array<Record<string, unknown>>;
  shadow_evidence: ShadowEvidence[];
  news: Array<Record<string, unknown>>;
}

export interface ResearchSnapshot {
  status: ApiEnvelope<StatusData>;
  overview: ApiEnvelope<OverviewData>;
  scanner: ApiEnvelope<ScannerData>;
}

export interface ResearchViewState {
  phase: "loading" | "ready";
  source: DataSource;
  snapshot: ResearchSnapshot | null;
  connectionMessage: string | null;
  loadedAt: string | null;
}
